File: frontend/src/api/sigmafi.ts
/**
 * SigmaFi Bond Protocol API
 *
 * TypeScript types and invoke wrappers for SigmaFi P2P bond protocol Tauri commands.
 */

import { invoke } from '@tauri-apps/api/core'


// =============================================================================
// Type Definitions
// =============================================================================

export interface CollateralToken {
  tokenId: string
  amount: number
  name: string | null
  decimals: number | null
}

export interface OpenOrder {
  boxId: string
  ergoTree: string
  creationHeight: number
  borrowerAddress: string
  loanTokenId: string
  loanTokenName: string
  loanTokenDecimals: number
  principal: number
  repayment: number
  maturityBlocks: number
  collateralErg: number
  collateralTokens: CollateralToken[]
  interestPercent: number
  apr: number
  collateralRatio: number | null
  isOwn: boolean
  transactionId: string
  outputIndex: number
}

export interface ActiveBond {
  boxId: string
  ergoTree: string
  originatingOrderId: string
  borrowerAddress: string
  lenderAddress: string
  loanTokenId: string
  loanTokenName: string
  loanTokenDecimals: number
  repayment: number
  maturityHeight: number
  collateralErg: number
  collateralTokens: CollateralToken[]
  blocksRemaining: number
  isLiquidable: boolean
  isRepayable: boolean
  isOwnLend: boolean
  isOwnBorrow: boolean
  transactionId: string
  outputIndex: number
}

export interface BondMarket {
  orders: OpenOrder[]
  bonds: ActiveBond[]
  blockHeight: number
}

export interface LoanToken {
  token_id: string
  name: string
  decimals: number
}

// =============================================================================
// API Functions
// =============================================================================

export async function fetchBondMarket(userAddress?: string): Promise<BondMarket> {
  return await invoke<BondMarket>('sigmafi_fetch_market', { userAddress: userAddress ?? null })
}

export async function getSupportedTokens(): Promise<LoanToken[]> {
  return await invoke<LoanToken[]>('sigmafi_get_tokens')
}

export async function buildOpenOrder(
  borrowerErgoTree: string,
  loanTokenId: string,
  principal: string,
  repayment: string,
  maturityBlocks: number,
  collateralErg: string,
  collateralTokensJson: string,
  userUtxos: object[],
  currentHeight: number,
): Promise<object> {
  return await invoke<object>('sigmafi_build_open_order', {
    borrowerErgoTree,
    loanTokenId,
    principal,
    repayment,
    maturityBlocks,
    collateralErg,
    collateralTokensJson,
    userUtxos,
    currentHeight,
  })
}

export async function buildCancelOrder(
  boxId: string,
  borrowerErgoTree: string,
  userUtxos: object[],
  currentHeight: number,
): Promise<object> {
  return await invoke<object>('sigmafi_build_cancel_order', {
    boxId,
    borrowerErgoTree,
    userUtxos,
    currentHeight,
  })
}

export async function buildCloseOrder(
  boxId: string,
  lenderErgoTree: string,
  uiFeeErgoTree: string,
  loanTokenId: string,
  userUtxos: object[],
  currentHeight: number,
): Promise<object> {
  return await invoke<object>('sigmafi_build_close_order', {
    boxId,
    lenderErgoTree,
    uiFeeErgoTree,
    loanTokenId,
    userUtxos,
    currentHeight,
  })
}

export async function buildRepay(
  boxId: string,
  loanTokenId: string,
  borrowerErgoTree: string,
  userUtxos: object[],
  currentHeight: number,
): Promise<object> {
  return await invoke<object>('sigmafi_build_repay', {
    boxId,
    loanTokenId,
    borrowerErgoTree,
    userUtxos,
    currentHeight,
  })
}

export async function buildLiquidate(
  boxId: string,
  lenderErgoTree: string,
  userUtxos: object[],
  currentHeight: number,
): Promise<object> {
  return await invoke<object>('sigmafi_build_liquidate', {
    boxId,
    lenderErgoTree,
    userUtxos,
    currentHeight,
  })
}

File: app/src/commands/sigmafi.rs
use citadel_api::services::sigmafi as sigmafi_svc;
use citadel_api::AppState;
use tauri::State;

/// Fetch the SigmaFi bond market (open orders + active bonds)
#[tauri::command]
pub async fn sigmafi_fetch_market(
    state: State<'_, AppState>,
    user_address: Option<String>,
) -> Result<sigmafi::BondMarket, String> {
    sigmafi_svc::fetch_market(&state, user_address.as_deref()).await
}

/// Get the supported loan tokens list
#[tauri::command]
pub async fn sigmafi_get_tokens() -> Result<serde_json::Value, String> {
    Ok(serde_json::Value::Array(sigmafi_svc::get_tokens()))
}

/// Build an open order transaction (borrower creates loan request)
#[tauri::command]
pub async fn sigmafi_build_open_order(
    borrower_ergo_tree: String,
    loan_token_id: String,
    principal: String,
    repayment: String,
    maturity_blocks: i32,
    collateral_erg: String,
    collateral_tokens_json: String,
    user_utxos: Vec<serde_json::Value>,
    current_height: i32,
) -> Result<serde_json::Value, String> {
    let parsed_utxos = super::parse_eip12_utxos(user_utxos)?;

    let principal: u64 = principal
        .parse()
        .map_err(|_| "Invalid principal amount".to_string())?;
    let repayment: u64 = repayment
        .parse()
        .map_err(|_| "Invalid repayment amount".to_string())?;
    let collateral_erg: u64 = collateral_erg
        .parse()
        .map_err(|_| "Invalid collateral ERG amount".to_string())?;

    let collateral_tokens: Vec<(String, u64)> =
        if collateral_tokens_json.is_empty() || collateral_tokens_json == "[]" {
            vec![]
        } else {
            serde_json::from_str(&collateral_tokens_json)
                .map_err(|e| format!("Invalid collateral tokens JSON: {}", e))?
        };

    let tx = sigmafi_svc::build_open_order(
        borrower_ergo_tree,
        loan_token_id,
        principal,
        repayment,
        maturity_blocks,
        collateral_erg,
        collateral_tokens,
        parsed_utxos,
        current_height,
    )?;
    serde_json::to_value(&tx).map_err(|e| format!("Failed to serialize tx: {}", e))
}

/// Build a cancel order transaction (borrower withdraws unfilled order)
#[tauri::command]
pub async fn sigmafi_build_cancel_order(
    state: State<'_, AppState>,
    box_id: String,
    borrower_ergo_tree: String,
    user_utxos: Vec<serde_json::Value>,
    current_height: i32,
) -> Result<serde_json::Value, String> {
    let parsed_utxos = super::parse_eip12_utxos(user_utxos)?;

    let tx = sigmafi_svc::build_cancel_order(
        &state,
        &box_id,
        borrower_ergo_tree,
        parsed_utxos,
        current_height,
    )
    .await?;
    serde_json::to_value(&tx).map_err(|e| format!("Failed to serialize tx: {}", e))
}

/// Build a close order transaction (lender fills an order, creating a bond)
#[tauri::command]
pub async fn sigmafi_build_close_order(
    state: State<'_, AppState>,
    box_id: String,
    lender_ergo_tree: String,
    ui_fee_ergo_tree: String,
    loan_token_id: String,
    user_utxos: Vec<serde_json::Value>,
    current_height: i32,
) -> Result<serde_json::Value, String> {
    let parsed_utxos = super::parse_eip12_utxos(user_utxos)?;

    let tx = sigmafi_svc::build_close_order(
        &state,
        &box_id,
        lender_ergo_tree,
        ui_fee_ergo_tree,
        loan_token_id,
        parsed_utxos,
        current_height,
    )
    .await?;
    serde_json::to_value(&tx).map_err(|e| format!("Failed to serialize tx: {}", e))
}

/// Build a repay transaction (borrower repays loan before maturity)
#[tauri::command]
pub async fn sigmafi_build_repay(
    state: State<'_, AppState>,
    box_id: String,
    loan_token_id: String,
    borrower_ergo_tree: String,
    user_utxos: Vec<serde_json::Value>,
    current_height: i32,
) -> Result<serde_json::Value, String> {
    let parsed_utxos = super::parse_eip12_utxos(user_utxos)?;

    let tx = sigmafi_svc::build_repay(
        &state,
        &box_id,
        loan_token_id,
        borrower_ergo_tree,
        parsed_utxos,
        current_height,
    )
    .await?;
    serde_json::to_value(&tx).map_err(|e| format!("Failed to serialize tx: {}", e))
}

/// Build a liquidate transaction (lender claims collateral after maturity)
#[tauri::command]
pub async fn sigmafi_build_liquidate(
    state: State<'_, AppState>,
    box_id: String,
    lender_ergo_tree: String,
    user_utxos: Vec<serde_json::Value>,
    current_height: i32,
) -> Result<serde_json::Value, String> {
    let parsed_utxos = super::parse_eip12_utxos(user_utxos)?;

    let tx = sigmafi_svc::build_liquidate(
        &state,
        &box_id,
        lender_ergo_tree,
        parsed_utxos,
        current_height,
    )
    .await?;
    serde_json::to_value(&tx).map_err(|e| format!("Failed to serialize tx: {}", e))
}
